discovery-provider/plugins/notifications/src/email/notifications/index.ts

```diff
@@ -343,21 +309,46 @@ export async function processEmailNotifications(
       days = 7
     }
     const startOffset = now.clone().subtract(days, 'days')
-    const pageCount = 1000
-    await getUsersCanNotify(identityDb, frequency, startOffset, pageCount, async (users) => {
-        if (Object.keys(users).length == 0) {
-          logger.info(`processEmailNotifications | No users to process. Exiting...`)
-          return
-        }
 
-        logger.info(
-          `processEmailNotifications | Beginning processing ${
-            Object.keys(users).length
-          } users at ${frequency} frequency`
-        )
+    // loop settings
+    const pageCount = remoteConfig.getFeatureVariableValue(NotificationsScheduledEmails, ScheduledEmailPluginMappings.PageCount, 1000)
+    let pageOffset = 0
+    let lastUser: number = 0;
+
+    // timeout settings since we run an infinite loop
+    const time = Date.now()
+    const timeoutMillis = 14400000
+    const timeout = time + timeoutMillis
 
-        await processGroupOfEmails(dnDb, identityDb, users, frequency, startOffset, remoteConfig)
-    })
+    while (true) {
+      const now = Date.now()
+      if (now > timeout) break
+      const userRows: { blockchainUserId: number; email: string }[] = await getUsersCanNotifyQuery(identityDb, startOffset, frequency, pageCount, lastUser)
+      pageOffset = pageOffset + pageCount
+      if (userRows.length === 0) break // once we've reached the end of users for this query
+      lastUser = userRows[userRows.length - 1].blockchainUserId
+      if (lastUser === undefined) {
```

Comment:
won't this always get hit before the check on line 334? if the length is 0 then lastUser will be undefined?

Reply:
that's a good point, should probably just combine those checks. good catch!